### PR TITLE
[Experimental Guns] No.6: Nikita

### PR DESCRIPTION
**Pull request by DeityLink**
Holy shit it's been an entire week since the previous one!? Ok then.

http://i.4cdn.org/vg/1438340075609.webm
- Experimental gun 6, the Nikita. A miniature cruise missile launcher. When you fire it, your camera will follow the missile and you'll control its orientation (you won't move yourself as long as you control the missile, which means as long as you keep the Nikita in your active hand). The missile moves quickly in straight lines and slows down when taking turns. It self detonates if over 100 tiles are crossed without bumping into anything. You can make it self detonate manually by clicking on the Nikita. EMPs disable the missile.
- Fixed the Syndie Rocker Launcher's fire sound. For some reason it was using the shotgun fire sound...

### CODE REVIEW

File: code/modules/mob/living/carbon/human/life.dm [hunk 1]
```diff
@@ -1623,7 +1623,7 @@ var/global/list/organ_damage_overlays = list(
 				if(temp_turf && (temp_turf.z != 1 && temp_turf.z != 5) || remoteview_target.stat!=CONSCIOUS)
 					src << "<span class='warning'>Your psy-connection grows too faint to maintain!</span>"
 					isRemoteObserve = 0
-			if(!isRemoteObserve && client && !client.adminobs && !iscamera(client.eye))
+			if(!isRemoteObserve && client && !client.adminobs && !iscamera(client.eye) && !istype(client.eye,/obj/item/projectile/nikita))
```
Comment (RemieRichards) [inline code/modules/mob/living/carbon/human/life.dm]: HNNNNNNNG

File: code/modules/mob/living/carbon/monkey/life.dm [hunk 2]
```diff
@@ -690,7 +690,7 @@
 			if (!( machine.check_eye(src) ))
 				reset_view(null)
 		else
-			if(client && !client.adminobs)
+			if(client && !client.adminobs && !istype(client.eye,/obj/item/projectile/nikita))
```
Comment (RemieRichards) [inline code/modules/mob/living/carbon/monkey/life.dm]: OH GOD

**Comment by clusterfack** [comment]
You should be able to do all this with just reset_view and relaymove without needing to put code anywhere else really

**Comment by dylanstrategie** [comment]
Why not inherit the rocket launcher at least for the basic behavior ? It's basically a guided missile rocket launcher

**Comment by ComicIronic** [comment]
SNNNNNNNNNNNNNNNNNNNNEEEEEEEEWWWWWWWWWWWFLLLLLLLLAAAAAAAAAAAAAAAAAAKKKKKKKEEEEEEEEEEEEE

**Comment by RemieRichards** [comment]
スノーフレーク

**Comment by Exxion** [comment]
What the *fuck* is that emag behavior in the webm? Does the rocket take you with it?
Why?

**Comment by Probe1** [comment]
hnnnnnnnnnnnnnnnnnnnng

**Comment by DeityLink** [comment]
@dylanstrategie it doesn't inherit the basic rocket launcher at all. They use a completely different type of ammo, which they process entirely differently, the projectile they fire behaves completely differently, I mean usually you make a weapon inherit from another when there's some code you want to re-use from it, but code-wise the Nikita isn't anything like the rocket launcher.

**Comment by DeityLink** [comment]
@Exxion an emagging doesn't necessarily make the item/machine better. Here the idea is to trick somebody else into using the nikita that you just emagged.

File: code/modules/mob/living/carbon/carbon.dm [hunk 3]
```diff
@@ -610,6 +610,13 @@
 
 	return null
 
+/mob/living/carbon/proc/isTeleViewing(var/client_eye)
```
Comment (ComicIronic) [inline code/modules/mob/living/carbon/carbon.dm]: Why is this a carbon proc

File: code/modules/projectiles/projectile/experimental.dm [hunk 4]
```diff
@@ -778,3 +778,139 @@
...
+			mob.client.perspective = EYE_PERSPECTIVE
+			mob.client.eye = src
+			mob.control_object = src
+			mob.canmove = 0
+
+	dir = get_dir(starting,original)
+	last_dir = dir
+
+	if(mob && emagged)
+		for(var/obj/item/W in mob)
```
Comment (ComicIronic) [inline code/modules/projectiles/projectile/experimental.dm]: in mob.get_all_slots()

File: code/modules/projectiles/guns/energy/experimental.dm [hunk 5]
```diff
@@ -506,3 +506,85 @@
...
+		return(BRUTELOSS)
+	else
+		user.visible_message("<span class='danger'>[user] fiddles with \the [src]'s safeties and suddenly aims it at \his feet! It looks like \he's trying to commit suicide.</span>")
+		spawn(10) //RUN YOU IDIOT, RUN
+			explosion(src.loc, -1, 1, 4, 8)
+			return(BRUTELOSS)
+	return
+
+/obj/item/weapon/gun/nikita/attackby(var/obj/item/A as obj, mob/user as mob)
+	if(istype(A, /obj/item/nikita))
```
Comment (ComicIronic) [inline code/modules/projectiles/guns/energy/experimental.dm]: You should be using the same system as all guns for things like this, not coding special cases.

File: code/modules/mob/mob_movement.dm [hunk 6]
```diff
@@ -187,6 +187,10 @@
 
 /client/proc/Move_object(direct)
 	if(mob && mob.control_object)
+		if(istype(mob.control_object,/obj/item/projectile/nikita))
```
Comment (ComicIronic) [inline code/modules/mob/mob_movement.dm]: Still has too much snewflake.

**Comment by clusterfack** [comment]
Can you make attack_self also detonate the missile, I don't think clicking on the missile itself is very intuitive and what people expect to detonate it

File: code/modules/mob/living/carbon/carbon.dm (hunk 3 above)
Comment (DeityLink) [inline code/modules/mob/living/carbon/carbon.dm]: so both monkeys and humans can use it.

File: code/modules/projectiles/guns/energy/experimental.dm (hunk 5 above)
Comment (DeityLink) [inline code/modules/projectiles/guns/energy/experimental.dm]: Other guns use ammo casings. But that doesn't fit missiles does it?

**Comment by DeityLink** [comment]
@clusterfack but that's what I did. Nikita is the name of the weapon.

File: code/modules/mob/mob_movement.dm (hunk 6 above)
Comment (DeityLink) [inline code/modules/mob/mob_movement.dm]: Move_object was previously only used for admins Possessing objects. I'd rather not code an entirely new proc and add new vars just for the nikita, now THAT would be snowflaky.

File: code/modules/mob/living/carbon/carbon.dm (hunk 3 above)
Comment (ComicIronic) [inline code/modules/mob/living/carbon/carbon.dm]: No, why is it a proc at the carbon level?
Why would only a carbon mob ever use this?

File: code/modules/projectiles/guns/energy/experimental.dm (hunk 5 above)
Comment (ComicIronic) [inline code/modules/projectiles/guns/energy/experimental.dm]: You can set the gun to not produce casings, in the same way the RPG does.
The point is that you should be using the general loading system, not coding your own.

File: code/modules/mob/mob_movement.dm (hunk 6 above)
Comment (ComicIronic) [inline code/modules/mob/mob_movement.dm]: No it would not be. You need to understand what snowflake means.

**Comment by DeityLink** [comment]
all good

**Comment by DeityLink** [comment]
Moved to GitLab: https://gitlab.com/vgstation/vgstation/merge_requests/2

**Pull request closed without merging**